File: cpp/930.binary-subarrays-with-sum.cpp
class Solution {
public:
  int numSubarraysWithSum(vector<int> &nums, int goal) {
    int i = 0, j = 0, n = nums.size(), res = 0, count = 0;

    for (; j < n; ++j) {
      if (nums[j] == 1) {
        goal--;
        count = 0;
      }

      while (goal == 0 && i <= j) {
        goal += nums[i];
        i++;
        count++;
        if (i > j - goal + 1) {
          break;
        }
      }

      while (goal < 0) {
        goal += nums[i];
        i++;
      }

      res += count;
    }

    return res;
  }
};

File: python/930.binary-subarrays-with-sum.py

from typing import List


class Solution:
    def numSubarraysWithSum(self, nums: List[int], goal: int) -> int:
        res = 0
        sum_cache = {0: 1}
        cur_sum = 0

        for num in nums:
            cur_sum += num
            res += sum_cache.get(cur_sum - goal, 0)
            sum_cache[cur_sum] = sum_cache.get(cur_sum, 0) + 1

        return res

